control.lua

```diff
@@ -1,366 +1,258 @@
   end
-  if ((color_type == "rainbow") or (color == "rainbow") or ((not color) and passengers_only)) then
-    color = make_rainbow(event_tick, train_id, settings, frequency, amplitude, center)
+  if ((mod_settings.color_type == "rainbow") or (color == "rainbow") or ((not color) and mod_settings.passengers_only)) then
+    color = make_rainbow(event_tick, train_id, mod_settings.frequency, mod_settings.amplitude, mod_settings.center)
   end
-  if color then
-    local position = stock.position
-    local surface = stock.surface
-    if sprite then
-      sprite = rendering.draw_sprite{
-        sprite = "train-trail",
-        target = position,
-        surface = surface,
-        tint = color,
-        x_scale = scale,
-        y_scale = scale,
-        render_layer = "radius-visualization",
-        time_to_live = length,
-      }
-    end
-    if light then
-      light = rendering.draw_light{
-        sprite = "train-trail",
-        target = position,
-        surface = surface,
-        color = color,
-        intensity = .175,
-        scale = scale * 1.75,
-        render_layer = "light-effect",
-        time_to_live = length,
-      }
-    end
+  if not color then return end
+  local position = stock.position
+  local surface = stock.surface
+  if mod_settings.sprite then
+    rendering.draw_sprite{
+      sprite = "train-trail",
+      target = position,
+      surface = surface,
+      tint = color,
+      x_scale = scale,
+      y_scale = scale,
+      render_layer = "radius-visualization",
+      time_to_live = length,
+    }
+  end
+  if mod_settings.light then
+    rendering.draw_light{
+      sprite = "train-trail",
+      target = position,
+      surface = surface,
+      color = color,
+      intensity = .175,
+      scale = scale * 1.75,
+      render_layer = "light-effect",
+      time_to_live = length,
+    }
   end
 end
 
 -- this one tries to reduce the weird ballooning and frying that happens when trains go really slowly, by making slower trains draw trails less frequently than faster ones
-local function draw_trails_based_on_speed(event, train, settings, sprite, light, color_override, length, scale, color_type, frequency, amplitude, center, passengers_only)
+---@param event_tick uint
+---@param mod_settings mod_settings
+---@param train LuaTrain
+---@param train_id uint
+local function draw_trails_based_on_speed(event_tick, mod_settings, train, train_id)
   local speed = train.speed
-  if not (speed == 0) then
-    local stock = false
-    if speed < 0 then
-      stock = train.back_stock
-    elseif speed > 0 then
-      stock = train.front_stock
-    end
-    if stock then
-      local event_tick = event.tick
-      local train_id = train.id
-      speed = speed * 216
-      --[[
-      216 is the conversion factor between tiles per tick and kilometers per hour; 60 * 3600 / 1000
-      --]]
+  if speed == 0 then return end
+  local stock = speed < 0 and train.back_stock or train.front_stock

```

Comment:
using the front/back stock is supposed to allow trails on trains that have a cargo wagon at the front, but i'm not confident about the speed being correlated to which one is leading